### PR TITLE
Complete Allblocks

### LINKED ISSUE #194: finalize allblocks.nim
a `allbocks.nim` file was introduced in #187 to document a new code block.
it should be:
- [ ] completed documenting all existing blocks
   - [ ] nbCodeInBlock
   - [ ] nbTextWithCode
- [ ] built when building docs (add a line in nimble file, see https://github.com/pietroppeter/nimib/blob/ce52a7613a4ac6ca5f80b5817abd018d919f745e/nimib.nimble#L33)
- [ ] linked in readme/index
- [ ] also a note should likely be added in CONTRIBUTING.md about its existence

### PR DESCRIPTION
**Pull request by dlesnoff**
Fix #194.
The `nbFile` and `nbShow` blocks are missing in the documentation but could be added after the PR.
I am not sure of their usage and was not very creative for examples today.

I added a TOC to list all blocks at the top and give quick access to their documentation.

The whale image (chosen for some kind of consistency with the icon) is licensed under [the Unsplash license](https://unsplash.com/license).

The documentation now requires Nimpy to work. which makes nimpy a *soft* dependency.

I removed the blocks description from the README.md.
The documentation link won't work until the docs are added to pietroppeter's blog.

Other partials (blocks too?) in `renders.nim` to add in `allblocks.nim`:
- [x] nbText
- [x] nbCode
- [x] nbFile
- [ ] nbJsFromCode
- [x] nimibCode

### CODE REVIEW

File: docsrc/allblocks.nim
Comment (HugoGranstrom) [inline docsrc/allblocks.nim:75]: This paragraph is incorrect. What you describe is string interpolation using the `fmt` macro which works with ordinary `nbText` equally well. `nbTextWithCode` differs from `nbText` only by the fact that the block also stores the code that produced the string passed to it. So visually and functionally there is no difference between `nbText` and `nbTextWithCode` for the average user.

File: docsrc/allblocks.nim
Comment (HugoGranstrom) [inline docsrc/allblocks.nim:98]: We are planning to move `nbPython` to nimibex (soon hopefully), so I don't know if we should document it in nimib. @pietroppeter any opinions?

File: docsrc/allblocks.nim
Comment (HugoGranstrom) [inline docsrc/allblocks.nim:32]: Since we (me and Pietro) use the fact that nimib can garantuee that your code compiles as one of the big arguments for using nimib, I think it would be worth it to add a big disclaimer on `nbCodeSkip` that makes that **very** clear. Something like:
```md
> `nbCodeSkip` does **not** garantuee in any way that the code that you show will compile nor run.
```

**Review by HugoGranstrom — CHANGES_REQUESTED** [review]
All in all a good summary of the blocks, thank you. Have have written a few comments on adding, changing and perhaps removing some parts though.

**Comment by HugoGranstrom** [comment]
>  nbText
> nbCode
> nbFile
> nbShow
> nbJsFromCode

All of these blocks should preferably be added, but that could be done in a follow-up PR by someone else as well :smile:

File: docsrc/allblocks.nim
Comment (pietroppeter) [reply docsrc/allblocks.nim:98]: I would skip it here

**Comment by pietroppeter** [comment]
> All of these blocks should preferably be added

I would skip `nbShow` which is not a NbBlock (it is more a command like nbInit) and should be deprecated (it is equivalent to nbSave with --nbShow runtime option). The others yeah should be added there.

**Comment by HugoGranstrom** [comment]
@pietroppeter `nbShow(df)` 😉

File: docsrc/allblocks.nim
Comment (dlesnoff) [reply docsrc/allblocks.nim:75]: When is the code stored, reused and not reused? When you recompile your code, does it know what it can reuse code?
If there is no difference for the average user, what is the point of this block specifically?
How would you explain this block then? May I ask you to write a suggested recommendation, so I can commit it directly with Github?

Comment (HugoGranstrom) [reply docsrc/allblocks.nim:75]: > When is the code stored, reused and not reused? When you recompile your code, does it know what it can reuse code? If there is no difference for the average user, what is the point of this block specifically?

Everything in nimib is a block, a call to `nbText` creates a new block containing the text and a call to `nbCode` creates a block containing the code and captured output. But what if you want to get the code passed to `nbText` as well and extract it? `nbText` doesn't store the code in its block (it did in the past but not anymore) so we created `nbTextWithCode` as an alternative when you want to not only get the string but also the code that created the string. The code `fmt"{1+1}"` produces the string `"2"`. `nbText` only gives us the `"2"` but `nbTextWithCode` stores the `fmt"{1+1}"` as well. The code can be accessed with `nb.blk.code` right after creating the `nbTextWithCode`. 

> How would you explain this block then? May I ask you to write a suggested recommendation, so I can commit it directly with Github?

Good question, I'll have to think about it. I'll make a suggestion tonight or tomorrow.

Comment (HugoGranstrom) [reply docsrc/allblocks.nim:75]:
```suggestion
`nbText` only stores the string it is given, but it doesn't store the code passed to `nbText`. For example, `nbText: fmt"{1+1}"` only stores the string `"2"` but not the code `fmt"{1+1}"` that produced that string. `nbTextWithCode` works like `nbText` but it also stores the code in the created block. It can be accessed with `nb.blk.code` right after the `nbTextWithCode` call. See the end of `numerical.nim` for an example.
```

Something like this, perhaps?

Comment (dlesnoff) [reply docsrc/allblocks.nim:75]: I commit it but I will move it at the bottom of the block's list.
It is more subtle than other blocks and will make a good transition to a "how to create a new block section".
I might detail one example of `nb.blk.code` for the explanation to be complete after.
Thank you for the explanation.

File: README.md
Comment (pietroppeter) [inline README.md:157]: as mentioned in the contributing.md, you need to change index.nim to modify the readme

File: docsrc/allblocks.nim
Comment (pietroppeter) [inline docsrc/allblocks.nim:29]:
```suggestion
nbCodeBlock: "nbCodeSkip"
```
looks a bit weird to have no space between `:` and the string (also in all other blocks below)

File: docsrc/allblocks.nim
Comment (pietroppeter) [inline docsrc/allblocks.nim:37]: here and for all the blocks we should probably use `nimibCode` to show the nimib code when calling the blocks (and we should document this usage of nimibCode).

File: docsrc/allblocks.nim
Comment (pietroppeter) [inline docsrc/allblocks.nim:119]: not just too long but also output you do not want to show (and maybe it is generated from a library), the important use case which should probably be cited explicitly is the case of ggplotnim, which writes to output everytime you save an image

**Review by pietroppeter — CHANGES_REQUESTED** [review]
thanks for working on this, I left a few comments to be addressed

File: docsrc/allblocks.nim
Comment (pietroppeter) [reply docsrc/allblocks.nim:32]: I am not sure I feel **that** strongly about this 😉, but more emphasis would not hurt :)

File: docsrc/allblocks.nim
Comment (dlesnoff) [reply docsrc/allblocks.nim:98]: Removed.

File: docsrc/allblocks.nim
Comment (dlesnoff) [reply docsrc/allblocks.nim:32]: I used a ⚠️ emoji for this.

File: docsrc/allblocks.nim
Comment (dlesnoff) [reply docsrc/allblocks.nim:37]: The distinction between the code of the block and the displayed block is confusing. Plus, it makes the code appears twice with not much addition (the title of the block).
A sentence separating both looks better to me. I still used nimibCode in my latest commits (to come) just so that you can compare.

File: docsrc/allblocks.nim
Comment (dlesnoff) [reply docsrc/allblocks.nim:119]: Fixed in a coming commit.

File: README.md
Comment (dlesnoff) [reply README.md:157]: I changed the index.nim file too.

**Comment by dlesnoff** [comment]
I merged `main` into my branch (integrating my 2 last PRs).
I added nbText, nbCode, nbFile, nimibCode blocks.
I replaced every markdown code with the nimibCode block.
I removed references to nbPython code block and the nimpy dependency.
Various fixes: 
 - Centered title example is smaller. It was too distracting.
 - Removed the incomplete document warning.
Please note that the nbFile example generates an extra file not even in the right directory.
It still lacks the nbJSFromCode block but I leave it to you a good JS example to illustrate it.
I did not understand, do we really drop the nbShow block when there is a special usage in one other document ?

I resolved all the conversations, made the requested changes and I am ready for another review.

File: README.md
Comment (dlesnoff) [reply README.md:157]: I made one extra change in the index.nim file though. I changed the title to
```
### other blocks
````

File: docsrc/allblocks.nim
Comment (dlesnoff) [inline docsrc/allblocks.nim:62]: I have to change the `is` elision

**Review by HugoGranstrom — CHANGES_REQUESTED** [review]
This looks great! :D I couldn't find many things to note this time around. Just make sure to add `allblocks.nim` to the docs and we will make sure the preview looks ok as well. And run `nimble readme` locally and push the updated `README.md`.

**Comment by dlesnoff** [comment]
I made the requested changes ;)

**Comment by pietroppeter** [comment]
Thanks looks very good! The link to see the all blocks preview is https://bc467ce645e09423fd70--nimib.netlify.app (the link we have is an absolute url - since it has to work also for readme -  so it does not work).
Very minor thing: you should use nimibCode also for nbCodeSkip

**Review by HugoGranstrom — APPROVED** [review]

**Comment by HugoGranstrom** [comment]
Once you fix Pietro's minor thing I'm content with merging this

**Comment by dlesnoff** [comment]
> The link to see the all blocks preview is https://bc467ce645e09423fd70--nimib.netlify.app/ (the link we have is an absolute url - since it has to work also for readme - so it does not work).

Sorry, I am not sure to understand this. Do you want me to change the link to the allblocks.nim document in the index.nim (and README.md) ?
This link does not point to an allblocks.nim but to the README.md of my PR.

**Comment by dlesnoff** [comment]
> you should use nimibCode also for nbCodeSkip

I fixed that! I am done with this PR.
To summarize, we should after this:
- [ ] reopen the issue https://github.com/pietroppeter/nimib/issues/194 for the two last blocks nbShow and nbJSFromCode.
- [ ] Add the allblocks document to pietroppeter.github.io website?
- [ ] Maybe make nimibCode separation between nimibCode rendered code block, the highlighted code and output of these code blocks more explicit? It would be cool to have the name of the files appearing in the code blocks (notebook.nim, notebook.html).

**Comment by pietroppeter** [comment]
> > The link to see the all blocks preview is https://bc467ce645e09423fd70--nimib.netlify.app/ (the link we have is an absolute url - since it has to work also for readme - so it does not work).
> 
> 
> 
> Sorry, I am not sure to understand this. Do you want me to change the link to the allblocks.nim document in the index.nim (and README.md) ?
> 
> This link does not point to an allblocks.nim but to the README.md of my PR.

Yeah, I forgot the important detail: https://b41374816c22d6cd819a--nimib.netlify.app/allblocks.html

Once you go to the link to latest netlify preview (and I updated with latest now), you can add allblocks.html to see the preview of that page.

**Comment by pietroppeter** [comment]
> > you should use nimibCode also for nbCodeSkip
> 
> 
> 
> I fixed that! I am done with this PR.

Thanks a lot, will merge this now!

> 
> To summarize, we should after this:
> 
> - [ ] reopen the issue https://github.com/pietroppeter/nimib/issues/194 for the two last blocks nbShow and nbJSFromCode.

👍 (nbJsFromCode has multiple variants, they are all already documented in interactivity.nim so not urgent)
> 
> - [ ] Add the allblocks document to pietroppeter.github.io website?

This should already happen automatically once this is merged

> - [ ] Maybe make nimibCode separation between nimibCode rendered code block, the highlighted code and output of these code blocks more explicit? It would be cool to have the name of the files appearing in the code blocks (notebook.nim, notebook.html).

Yeah, we could customize partial of nimibCode for this document in order to have something like:

⬇️⬇️⬇️ source
(Here the source code)
⬇️⬇️⬇️ output
(Here the output)

**Pull request merged**
